universe/Fleet.cpp

```diff
@@ -79,6 +79,34 @@ namespace {
             system->Insert(ship);
         }
     }
+
+    /** Return \p full_route terminates at \p last_system or before the first
+        system not known to the \p empire_id. */
+    std::list<int> TruncateRouteToEndAtSystem(const std::list<int>& full_route, int empire_id, int last_system) {
+
+        if (full_route.empty() || (last_system == INVALID_OBJECT_ID))
+            return std::list<int>();
+
+        auto visible_end_it = full_route.cend();
+        if (last_system != full_route.back()) {
+            visible_end_it = std::find(full_route.begin(), full_route.end(), last_system);
+
+            // if requested last system not in route, do nothing
+            if (visible_end_it == full_route.end())
+                return std::list<int>();
+
+            ++visible_end_it;
+        }
+
+        // remove any extra systems from the route after the apparent destination
+        auto end_it = std::find_if(full_route.begin(), visible_end_it,
+                                   boost::bind(&SystemHasNoVisibleStarlanes, _1, empire_id));
```

Comment:
this is a bit misleading due to that function's name... maybe add a comment that this also excludes systems not known to the specified empire?

Reply:
also maybe a comment noting that if a ship is in a system, or is moving between two systems, then those systems should always be visible to the empire that owns the ship...

Reply:
I expanded the comment.